types.go

```diff
@@ -97,52 +97,44 @@ type Stats struct {
 	Runtime         time.Duration // Total runtime of the optimization
 }
 
-// FunctionInfo is data to give to the optimizer about the objective function.
-type FunctionInfo struct {
-	IsGradient                bool
-	IsHessian                 bool
-	IsFunctionGradient        bool
-	IsFunctionGradientHessian bool
-	IsStatuser                bool
+// ProblemInfo is data to give to the optimizer about the objective function.
+type ProblemInfo struct {
+	HasGradient                bool
+	HasHessian                 bool
+	HasFunctionGradient        bool
+	HasFunctionGradientHessian bool
+	HasStatus                  bool
 }
 
-// functionInfo contains information about which interfaces the objective
-// function F implements and the actual methods of F that have been
-// successfully type switched.
-type functionInfo struct {
-	FunctionInfo
-
-	function Function
-	gradient Gradient
-	hessian  Hessian
-	statuser Statuser
+// problemInfo contains information about which methods the optimization
+// problem provides as well as the optimization problem itself.
+type problemInfo struct {
```

Comment:
I think we should remove this type. It existed to avoid repeated type assertions on `Function`, now we can easily test fields of `Problem` for nil. Internal functions that take `problemInfo` can take `Problem` directly. `satisfies` can be an unexported method on `Problem` until we make it exported or find a better name and place for it. The other PR in queue removes passing of `ProblemInfo` to the `Recorder`, so after it is merged `problemInfo` won't be needed in `Local()`.